Enroll new apps in decrypted diffs of credentials

jonathanhefner opened the pull request:
Users, especially those who are creating an app for the first time, might not be aware of the `rails credentials:diff --enroll` command.  This commit enrolls new apps in decrypted diffs by default so that users benefit automatically.  This behavior can be opted out of with the `--skip-decrypted-diffs` flag.

jonathanhefner committed:
Enroll new apps in decrypted diffs of credentials

Users, especially those who are creating an app for the first time,
might not be aware of the `rails credentials:diff --enroll` command.
This commit enrolls new apps in decrypted diffs by default so that users
benefit automatically.  This behavior can be opted out of with the app
generator's `--skip-decrypted-diffs` flag.

jonathanhefner commented:
I will come back to this.  I'm going to think about how best to respect the `--quiet` flag.

jonathanhefner commented:
If erikhuda/thor#729 is merged (and released), the [credentials command call](https://github.com/rails/rails/pull/39686/files#diff-248322211d56bb3b2ab59ac68a2839f5R198) should respect the `--quiet` flag just as the app generator does.

If waiting for erikhuda/thor#729 isn't feasible, I can explore more direct solutions.

rafaelfranca commented:
What is the requirement for the `quiet` option? The diff doesn't use that option.

jonathanhefner commented:
@rafaelfranca The `--quiet` option is for the app generator, which eventually invokes [`enroll_project_in_credentials_diffing`](https://github.com/rails/rails/blob/4ed00c4d9a871638a518b91a12a3c17767bfd1b4/railties/lib/rails/commands/credentials/credentials_command/diffing.rb#L4-L16).  I can investigate possible workarounds, but erikhuda/thor#729 seemed like the cleanest way to make those `say` calls silent when necessary.

rafaelfranca commented:
Ah I see. So we don't need to wait a release of Thor right? I can do a release, but I was wondering if we can apply those changes without a release and bumping the minimal version of thor.

kaspth reviewed (approved):
Awesome! I think we'll need another look at the diffing command too. iirc, there were inputs it wasn't doing so well with, but unfortunately I don't remember what they were right now.

dhh commented:
This is creating a lot of noise, and outputting the secret_key_base for production, straight into the console on a new app. That doesn't seem right?

```
      create  config/master.key
      append  .gitignore
Adding config/credentials.yml.enc to store encrypted credentials.

The following content has been encrypted with the Rails master key:

# aws:
#   access_key_id: 123
#   secret_access_key: 345

# Used as the base secret for all MessageVerifiers in Rails, including the one protecting cookies.
secret_key_base: aa1a8fc23a37298a0469109a3f3fb14f4428adbac236f535c5f1db5c6e3f91c6db603b004fa113359a0cdd97385ad4866f9d8d39920a98

You can edit encrypted credentials with `bin/rails credentials:edit`.

       rails  credentials:diff --enroll
Enrolled project in credentials file diffing!
Rails ensures the rails_credentials diff driver is set when running `bin/rails credentials:edit`. See `bin/rails credentials:help` for more.
      create  config/boot.rb
      create  config/database.yml
```

jonathanhefner commented:
@dhh The `config/credentials.yml.enc` dump is due to #45543.  I had changed the logging to depend on Thor's `options[:quiet]`, but neglected to set `quiet: true` in the application generator.  I've fixed that in #46766.

The

```
Enrolled project in credentials file diffing!
Rails ensures the rails_credentials diff driver is set when running `bin/rails credentials:edit`. See `bin/rails credentials:help` for more.
```

output is due to this PR though.  My thinking was that drawing attention to the feature would help the user later recognize why `git diff` shows a decrypted diff.  But showing a reminder on the first `bin/rails credentials:edit` would probably be more helpful, so I've done that in #46767.
